gcsmigrate/shared.py

```diff
@@ -0,0 +1,354 @@
+
+    def save_to_firestore(self, collection, md5, gcs_bucket):
+        """Save record to firestore certifying successful migration."""
+        doc_ref = collection.document(str(self.fid))
+        return doc_ref.set(
+            {"fid": self.fid, "sha1": self.sha1sum, "md5": md5, "bucket": gcs_bucket}
+        )
+
+    def to_json(self):
+        """Serialize as JSON."""
+        return bytes(
+            json.dumps({"length": self.length, "fid": self.fid, "dkey": self.dkey}),
+            "utf-8",
+        )
+
+    @classmethod
+    def from_json(cls, json_str):
+        """Create MogileFile from JSON string."""
+        return MogileFile(**json.loads(json_str))
+
+    def verify(self, fid, md5, sha1, bucket, bucket_map, gcs_client):
+        """Verify (with assert) this mogile file against asserted values."""
+        print(f"Verifying {self.fid}")
+        assert self.fid == fid, f"fid {self.fid} not migrated"
+        new_bucket = bucket_map[self.mogile_bucket]
+        assert new_bucket == bucket
+        remote_md5 = self.contentrepo_exists_in_bucket(gcs_client, new_bucket)
+        assert remote_md5 == md5, f"{self.fid} has wrong MD5 sum"
+        assert self.sha1sum == sha1, f"{self.fid} has wrong SHA1 sum"
+
+
+def get_mogile_files_from_database(
+    database_url, limit=None, fids=None, excluded_fids=set()
+):
+    """Return a generator for all mogile files in the database."""
+    config = dj_database_url.parse(database_url)
+    connection = pymysql.connect(
+        host=config["HOST"],
+        user=config["USER"],
+        password=config["PASSWORD"],
+        db=config["NAME"],
+        cursorclass=pymysql.cursors.SSCursor,
+    )
+
+    try:
+        cursor = connection.cursor()
+        if limit is not None:
+            sql = f"SELECT * FROM file LIMIT {limit}"
+        elif fids is not None:
+            fids_in = ", ".join(fids)
+            sql = f"SELECT * FROM file WHERE fid IN ({fids_in})"
+        else:
+            sql = "SELECT * FROM file"
+        cursor.execute(sql)
+        row = cursor.fetchone()
+        while row:
+            if row[0] not in excluded_fids:
+                yield MogileFile.parse_row(row)
+            row = cursor.fetchone()
+    finally:
+        connection.close()
+
+
+def make_generator_from_args(args):
+    """Make a mogile_file generator from args.
+
+    Args are either a list of fids to process or a single file that
+    contains a list of fids to exclude.
+    """
+    fids = None
+    excluded_fids = set()
+    if len(args) > 0:
+        if args[0].isdigit():
+            fids = args
+        else:
+            with open(args[0]) as f:
+                for line in f:
+                    excluded_fids.add(int(line))
+            print(f"Excluding {len(excluded_fids)} fids.")
+    return get_mogile_files_from_database(
+        os.environ["MOGILE_DATABASE_URL"], fids=fids, excluded_fids=excluded_fids
+    )
+
+
+def chunked(iterable, size):
```

Comment:
no need to change, but itertools has a nice recipe for this which returns an iterable:

```py
from itertools import zip_longest

def grouper(iterable, n, fillvalue=None):
    "Collect data into fixed-length chunks or blocks"
    # grouper('ABCDEFG', 3, 'x') --> ABC DEF Gxx"
    args = [iter(iterable)] * n
    return zip_longest(*args, fillvalue=fillvalue)
```

(from https://docs.python.org/3/library/itertools.html#itertools-recipes )

Reply:
This was dead code, sorry! I delete it. Also, good to know! I thought Iooked. :flushed: